### ISSUE TITLE
Error when querying chaincode - Causing problems processing transactions

### ISSUE THREAD

**Issue by johnwhitton** [hyperledger/composer#1783]
## Context

We are replicating transactions on to Fabric using the Composer rest server.
As part of this process we run a query every 5 seconds to ensure that the Fabric Rest Server is up.
After a period of time we are getting errors when querying the Composer rest server and when trying to process transactions we get failures.

## Expected Behavior

Query should respond without getting an error

## Actual Behavior

We get an error when querying the chaincode, below are the logs from the fabric composer rest server.

```
0|composer | WARNING: NODE_APP_INSTANCE value of '0' did not match any instance config file names.
0|composer | WARNING: See https://github.com/lorenwest/node-config/wiki/Strict-Mode
0|composer | Discovering types from business network definition ...
0|composer | D0807 20:58:10.660643200    1056 env_linux.c:66]             Warning: insecure environment read function 'getenv' used
0|composer | D0807 20:58:11.023339434    1056 env_linux.c:66]             Warning: insecure environment read function 'getenv' used
0|composer | Connection fails: Error: error trying to ping. Error: error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer | It will be retried for the next request.
0|composer | Exception: Error: error trying to ping. Error: error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer | Error: error trying to ping. Error: error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer |     at queryChainCode.then.catch (/usr/local/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:707:34)
PM2        | App [composer-rest-server] with id [0] and pid [1056], exited with code [1] via signal [SIGINT]
PM2        | Starting execution sequence in -fork mode- for app name:composer-rest-server id:0
PM2        | App name:composer-rest-server id:0 online
```

## Steps to Reproduce

1.
2.
3.
4.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [ ] [GitHub Issues](https://github.com/hyperledger/composer/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

## Context

This leaves us unable to replicate transactions on chain.
Rebooting both the peer and the composer rest server usually provides a temporary solution

## Your Environment

* Version used: composer-rest-server:0.10.0

**Comment by davidkel** [comment]
@johnwhitton could you get the chaincode logs and paste them here please ? the command is
`docker logs {container_name} 2>cclogs.txt`
Where {container_name} will be a container starting with the text `dev-` and will have the name of you business network as part of it.

**Comment by johnwhitton** [comment]
@davidkel Thanks for the quick response, attached you'll find the log file 
[peer.txt](https://github.com/hyperledger/composer/files/1208207/peer.txt)
If I can provide more information  you can also reach me on rocketchat as johnwhitton
Here is the link to the question I raised on rocketchat
https://chat.hyperledger.org/channel/composer?msg=kDzZmbAuQ4WTuHuZL

**Please note this is impacting the ability for us to go live with a client we are working with**

**Comment by johnwhitton** [comment]
Attaching the chaincode logs
[dev-peer0-log.txt](https://github.com/hyperledger/composer/files/1208285/dev-peer0-log.txt)

**Comment by johnwhitton** [comment]
And the debug log
[dev-peer0-log-d2.txt](https://github.com/hyperledger/composer/files/1208406/dev-peer0-log-d2.txt)

**Comment by johnwhitton** [comment]
Chatted with @davidkel on Rocketchat and he suggested that this might be related to ticket https://github.com/hyperledger/composer/issues/1670
And recommended that we moved to a later version of [composer-rest-server](https://hub.docker.com/r/hyperledger/composer-rest-server/tags/)
We'll upgrade to 0.11.2 and see if this resolves the issue.
We'll also look at our health checks and see if we can do this more effectively as the current mechanism was passing an invalid request and compounding the issue.

**Comment by RashitKhamidullin** [comment]
It looks like I have encountered the same problem. 
@johnwhitton have you solved the issue?

**Comment by johnwhitton** [comment]
HI @RashitKhamidullin I was not able to reproduce this with debugging on and so have not shared the logs requested. The current assumption is that once we run out of memory than the valstack error occurs. We have upgraded our composer-rest-server to 0.11.2 as suggested above. This appears to have addressed the memory leak issue. We tested this for a period of 8 hours running transactions and used docker stats to see the memory usage for the new rest server container. The newer version did not appear to have the memory leak issue. I'm going to close this ticket as we believe the upgrade resolved the issue and if we see the error again, I'll raise a new ticket and link it to this one.

**Comment by johnwhitton** [comment]
Hi all,

Reopening this, as we have run into the same error using Composer 0.11.2 however this may relate to an problem with the underlying peer.
Below is some log info

**Composer Rest Server**
```
PM2        | App [composer-rest-server] with id [0] and pid [8595], exited with code [1] via signal [SIGINT]
PM2        | Starting execution sequence in -fork mode- for app name:composer-rest-server id:0
PM2        | App name:composer-rest-server id:0 online
0|composer | WARNING: NODE_APP_INSTANCE value of '0' did not match any instance config file names.
0|composer | WARNING: See https://github.com/lorenwest/node-config/wiki/Strict-Mode
0|composer | Discovering types from business network definition ...
0|composer | D0828 15:51:44.842150412    8607 env_linux.c:66]             Warning: insecure environment read function 'getenv' used
0|composer | D0828 15:51:45.253113275    8607 env_linux.c:66]             Warning: insecure environment read function 'getenv' used
0|composer | Connection fails: Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer | It will be retried for the next request.
0|composer | Exception: Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer | Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer |     at channel.queryByChaincode.then.catch (/usr/local/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:782:34)
```

**Peer**
```
2017-08-28 15:59:09.514 UTC [endorser] simulateProposal -> DEBU 440e3c Exit
2017-08-28 15:59:09.514 UTC [endorser] ProcessProposal -> ERRO 440e3d simulateProposal() resulted in chaincode response status 500 for txid: 149c4ebd7b0f2c7496fb4cfcb39c82567661a589f5b8ab16ef6fdb11d4392629
2017-08-28 15:59:09.514 UTC [lockbasedtxmgr] Done -> DEBU 440e3e Done with transaction simulation / query execution [5ff99b1e-143c-4991-ba23-cb98ee09762d]
```

Note : We got this error when deploying a new network which has a health check of a system ping every 15 seconds. (i.e. our application calls GET /system/ping/ every 15 seconds to ensure Fabric is up.

**Comment by davidkel** [comment]
@johnwhitton unfortunately there isn't anything in the above logs that gives any useful information. Can you provide some more information
- how soon does the problem occur once the fabric has been started (as opposed to the rest server)
- what are the memory stats of the chaincode container
- can you obtain the chaincode container logs when the problem occurs

**Comment by johnwhitton** [comment]
Hi,

Here is some additional information. @davidkel 

**Background**
- We have deployed a business network which has amongst others an asset balance
- We run a "health check" which effectively gets a balance by it's UUID
- If we run this for a number of days we start to see 
```
0|composer | Unhandled error for request GET /api/Balance?filter=%7B%22where%22%3A%7B%22uuid%22%3A%22test_uuid%22%7D%7D: Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer |     at channel.queryByChaincode.then.catch (/usr/local/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:782:34)
```
- If we restart the peer and composer-rest-server (which effectivley restarts the dev-peer0) the issue goes away
- We are running composer rest server 0.11.2
- We are running fabric x86_64-1.0.1
- We are running dockerized containers on Google Cloud Platform
- We are using a cluster which has 1 VM with 2cpu's and 7.5 GB of RAM

**Sample Error**
- Curl
`curl -X GET --header 'Accept: application/json' 'http://104.198.65.29:3000/api/Balance?filter=%7B%22where%22%3A%7B%22uuid%22%3A%20%22test_uuid%22%7D%7D'`
- Response
```
{
  "error": {
    "statusCode": 500,
    "name": "Error",
    "message": "Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)",
    "stack": "Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)\n    at channel.queryByChaincode.then.catch (/usr/local/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:782:34)"
  }
}
```

**Hypothesis**
- There is a slow memory leak causing the issue

**Additional Information**
_Note : attached are logs for all of the below_

- **Composer Rest Server**
  - The composer rest server is echoing a message provided by dev_peer
```
0|composer | Unhandled error for request GET /api/Balance?filter=%7B%22where%22%3A%7B%22uuid%22%3A%22test_uuid%22%7D%7D: Error: Error trying to query chaincode. Error: chaincode error (status: 500, message: RangeError: valstack limit)
0|composer |     at channel.queryByChaincode.then.catch (/usr/local/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:782:34)
```
- **Channel (dev_peer_0)**
  - Identity Manager is getting a Unhandled Promise Rejection
```
2017-09-02 12:07:51.140 UTC [Composer] Error -> ERRO 890d @JS : IdentityManager          :getIdentity()             !! getIdentity, [object Object]
Error: Unhandled promise rejection {identityName:"admin"}
```
  - Memory usage continues to grow as observed by docker stats
```
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
2007ac590c3f        2.79%               533.7 MB / 2.147 GB   24.85%              0 B / 0 B           0 B / 0 B           8
```

- **Peer**
  - The peer is giving a status 500 error 
```
[31m2017-09-02 17:07:50.013 UTC [endorser] ProcessProposal -> ERRO 7b2469[0m simulateProposal() resulted in chaincode response status 500 for txid: c304acfa51f28418136d67ecf5e1299215be508da0d104b50e727d1e93f4f3ec
```
- **Certification Authority**
  - Does not appear to be getting any identity requests associated with this 

- **Orderer**
  - Order logs attached - seems to be creating a new identity instance but may have permissioning issue


**Logs**

[orderer_log.txt](https://github.com/hyperledger/composer/files/1272365/orderer_log.txt)
[dev_peer_log.txt](https://github.com/hyperledger/composer/files/1272366/dev_peer_log.txt)
[ca_log.txt](https://github.com/hyperledger/composer/files/1272369/ca_log.txt)
[peer_log.txt](https://github.com/hyperledger/composer/files/1272368/peer_log.txt)
[composer-rest-server_log.txt](https://github.com/hyperledger/composer/files/1272367/composer-rest-server_log.txt)

**Comment by muralisrini** [comment]
@davidkel @johnwhitton the simulateProposal 500 error is coming from the chaincode .  If you add CORE_VM_DOCKER_ATTACHSTDOUT=true to the peer we can capture any output from the chaincode (hopefully some error will be output) which will help debug.

**Comment by johnwhitton** [comment]
@muralisrini 
Let me know if this log has the additional information
[peer_debug_log.txt](https://github.com/hyperledger/composer/files/1272575/peer_debug_log.txt)

**Comment by muralisrini** [comment]
@johnwhitton The new information shows a GetState call not finding a key. This needs some with chaincode (and possibly composer) knowledge to debug to know if this be the cause of the error.

[36m2017-09-02 22:41:05.100 UTC [chaincode] func1 -> DEBU 1fee2[0m [614986eb] getting state for chaincode loyyal-network, key Asset:com.loyyal.goyyal.Balancetest_uuid, channel loyyalchannel
[36m2017-09-02 22:41:05.100 UTC [stateleveldb] GetState -> DEBU 1fee3[0m GetState(). ns=loyyal-network, key=Asset:com.loyyal.goyyal.Balancetest_uuid
### **[36m2017-09-02 22:41:05.100 UTC [chaincode] func1 -> DEBU 1fee4[0m [614986eb]No state associated with key: Asset:com.loyyal.goyyal.Balancetest_uuid. Sending RESPONSE with an empty payload**
[36m2017-09-02 22:41:05.100 UTC [chaincode] 1 -> DEBU 1fee5[0m [614986eb]handleGetState serial send RESPONSE
[36m2017-09-02 22:41:05.103 UTC [chaincode] processStream -> DEBU 1fee6[0m [614986eb]Received message COMPLETED from shim
[36m2017-09-02 22:41:05.103 UTC [chaincode] HandleMessage -> DEBU 1fee7[0m [614986eb]Fabric side Handling ChaincodeMessage of type: COMPLETED in state ready
[36m2017-09-02 22:41:05.103 UTC [chaincode] HandleMessage -> DEBU 1fee8[0m [614986eb5cbcd9f7c8c2be61d95852a74cd5347fd03fae3e829669445670a81b]HandleMessage- COMPLETED. Notify
[36m2017-09-02 22:41:05.103 UTC [chaincode] notify -> DEBU 1fee9[0m notifying Txid:614986eb5cbcd9f7c8c2be61d95852a74cd5347fd03fae3e829669445670a81b
[36m2017-09-02 22:41:05.103 UTC [chaincode] Execute -> DEBU 1feea[0m Exit
[36m2017-09-02 22:41:05.103 UTC [endorser] callChaincode -> DEBU 1feeb[0m Exit
[36m2017-09-02 22:41:05.103 UTC [lockbasedtxmgr] GetTxSimulationResults -> DEBU 1feec[0m Simulation completed, getting simulation results
[36m2017-09-02 22:41:05.103 UTC [lockbasedtxmgr] Done -> DEBU 1feed[0m Done with transaction simulation / query execution [01192b2c-f7db-4445-89ce-6dad8157dcae]
[36m2017-09-02 22:41:05.103 UTC [endorser] simulateProposal -> DEBU 1feee[0m Exit
[31m2017-09-02 22:41:05.103 UTC [endorser] ProcessProposal -> ERRO 1feef[0m simulateProposal() resulted in chaincode response status 500 for txid: 614986eb5cbcd9f7c8c2be61d95852a74cd5347fd03fae3e829669445670a81b

**Comment by johnwhitton** [comment]
@muralisrini My assumption is that I'm trying to get an asset that does not exist.
As you can see from the curl command above. Where looking for a balance with an id of test_uuid.
There is no asset with that id, hence the empty response.
@davidkel - Maybe you can add more insight.

**Comment by muralisrini** [comment]
@johnwhitton To be clear, just fishing here and only thing I see is that "not found" message followed by the 500 error. As you point out, this might be an expected result and something else is going on. We need more debug between "not found" and the 500 error.

**Comment by davidkel** [comment]
@johnwhitton This problem is related to the javascript engine we are currently using (duktape) and although I may have an idea that might help with this specific problem I don't know if it will actually fix it without being able to recreate. At the moment we are focusing our efforts on making node.js our javascript engine and getting rid of duktape altogether (as we are seeing other issues with the implementation as well) so for now we aren't able to look any further into this. There will be lots of advantages to using node.js including up to date language features, being able to use npm modules, smaller memory footprint etc and see that this will be a very exciting capability for composer.

**Comment by johnwhitton** [comment]
Thanks @davidkel is there a ticket which I can use to monitor the progress of the migration to node.js from duktape.

**Comment by davidkel** [comment]
@johnwhitton issue is https://github.com/hyperledger/composer/issues/61

**Comment by mamtabhardwaj12** [comment]
Hello Johnwhitton,

I am running composer rest server,i have uploaded my bna file on playground it works properly for some times but after that i got error, so that i am not able to add any participaints.
Thank you.
![composererror](https://user-images.githubusercontent.com/27012406/30843026-1acd77b6-a2a4-11e7-9dfa-4a63add140c4.png)

**Comment by cpagravel** [comment]
I am also getting the valstack error. I am using Composer REST Server HyperLedger Fabric generated from https://github.com/IBM-Blockchain/ibm-container-service 

I followed the steps described here: https://ibm-blockchain.github.io/

It's the second time it's happened. The first time, I ended up removing and starting the entire blockchain from scratch. I fear that is what I will have to do again.

This error seems to pop up when the block chain gets too large.

**Comment by davidkel** [comment]
Big improvements have been made to issues around the duktape implementation, so hopefully these problems have been fixed and you should move to the 0.16.x release path.
Also with fabric 1.1 now coming out we have dropped duktape completely in our 0.17.x release path and now use the built in node support in fabric 1.1 and this provides a big improvement in performance over the 16.x.
